keanu-project/src/main/java/io/improbable/keanu/tensor/TensorShapeValidation.java

```diff
@@ -90,4 +93,14 @@ public static void checkTensorsMatchNonScalarShapeOrAreScalar(int[] proposalShap
         return uniqueShapes.iterator().next().getShape();
     }
 
+    public static int[] checkShapesCanBeConcatenated(int dimension, int[]... shapes) {
```

Comment:
Bug here with calculating the placeholder. Wrote something that might work for you....
```
    public static int[] checkShapesCanBeConcatenated(int dimension, int[]... shapes) {

        int[] concatShape = Arrays.copyOf(shapes[0], shapes[0].length);
        for (int i = 1; i < shapes.length; i++) {
            if (shapes[i].length != concatShape.length) {
                throw new IllegalArgumentException("Cannot concat shapes of different ranks");
            }

            for (int dim = 0; dim < shapes[i].length; dim++) {
                if (dim == dimension) {
                    concatShape[dim] += shapes[i][dim];
                } else {
                    if (shapes[i][dim] != concatShape[dim]) {
                        throw new IllegalArgumentException("Cannot concat mismatched shapes");
                    }
                }
            }
        }
        return concatShape;
    }
```

Reply:
Nice - changed